provider/cloudflare/cloudflare.go

```diff
@@ -394,34 +385,37 @@ func (p *CloudFlareProvider) submitChanges(ctx context.Context, changes []*cloud
+			}
+			chs, chErr := p.listCustomHostnamesWithPagination(ctx, zoneID)
+			if chErr != nil {
+				return fmt.Errorf("could not fetch custom hostnames from zone, %v", chErr)
+			}
 			if change.Action == cloudFlareUpdate {
 				if recordTypeCustomHostnameSupported[change.ResourceRecord.Type] {
-					chID, oldCh := p.getCustomHostnameIDbyOrigin(chs, change.ResourceRecord.Name)
-					if chID == "" && change.CustomHostname.Hostname != "" {
-						log.WithFields(logFields).Infof("Adding custom hostname %v", change.CustomHostname.Hostname)
-						_, chErr := p.Client.CreateCustomHostname(ctx, zoneID, change.CustomHostname)
-						if chErr != nil {
-							failedChange = true
-							log.WithFields(logFields).Errorf("failed to add custom hostname %v: %v", change.CustomHostname.Hostname, chErr)
-						}
-					} else if chID != "" && oldCh != "" && change.CustomHostname.Hostname == "" {
-						log.WithFields(logFields).Infof("Removing custom hostname %v", change.CustomHostname.Hostname)
-						chErr := p.Client.DeleteCustomHostname(ctx, zoneID, chID)
-						if chErr != nil {
-							failedChange = true
-							log.WithFields(logFields).Errorf("failed to remove custom hostname %v: %v", change.CustomHostname.Hostname, chErr)
+					prevCh := change.CustomHostnamePrev
+					newCh := change.CustomHostname.Hostname
+					if prevCh != "" {
+						prevChID, _ := p.getCustomHostnameOrigin(chs, prevCh)
+						if prevChID != "" && prevCh != newCh {
+							log.WithFields(logFields).Infof("Removing previous custom hostname %v/%v", prevChID, prevCh)
+							chErr := p.Client.DeleteCustomHostname(ctx, zoneID, prevChID)
+							if chErr != nil {
+								failedChange = true
+								log.WithFields(logFields).Errorf("failed to remove previous custom hostname %v/%v: %v", prevChID, prevCh, chErr)
+							}
 						}
-					} else if chID != "" && change.CustomHostname.Hostname != "" && oldCh != change.CustomHostname.Hostname {
-						log.WithFields(logFields).Infof("Replacing custom hostname: %v/%v to %v", chID, oldCh, change.CustomHostname.Hostname)
-						chDelErr := p.Client.DeleteCustomHostname(ctx, zoneID, chID)
-						if chDelErr != nil {
-							failedChange = true
-							log.WithFields(logFields).Errorf("failed to remove replacing custom hostname %v/%v: %v", chID, oldCh, chDelErr)
-						}
-						_, chAddErr := p.Client.CreateCustomHostname(ctx, zoneID, change.CustomHostname)
-						if chAddErr != nil {
-							failedChange = true
-							log.WithFields(logFields).Errorf("failed to add replacing custom hostname %v: %v", change.CustomHostname.Hostname, chAddErr)
+					}
+					if newCh != "" {
+						if prevCh != newCh {
+							log.WithFields(logFields).Infof("Adding custom hostname %v", newCh)
+							_, chErr := p.Client.CreateCustomHostname(ctx, zoneID, change.CustomHostname)
+							if chErr != nil {
+								failedChange = true
+								log.WithFields(logFields).Errorf("failed to add custom hostname %v: %v", newCh, chErr)
+							}
```

Comment:
not covered with unit tests

![Screenshot 2025-03-06 at 07 47 13](https://github.com/user-attachments/assets/2d663f98-63bc-44b8-b150-12445eb58483)

Reply:
covered